File: java/simpleregistration/SimpleRegistration.java
/**
 * @file
 * Really simple product registration.
 * 
 * Simple registration tool, validates a username and entry code for a user,
 * returns whether it is valid.
 * 
 * Keeping honest people honest :)
 * 
 * @package simpleregistration
 * @copyright Marcus Povey Consulting, All rights reserved.
 * @license The MIT License (see LICENCE.txt), other licenses available.
 * @author Marcus Povey <marcus@marcus-povey.co.uk>
 * @link http://www.marcus-povey.co.uk
 */
package simpleregistration;


/**
 * @author Marcus Povey <marcus@marcus-povey.co.uk>
 */
public class SimpleRegistration {
    
    private static String HASHTABLE = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
   
    /**
     * The salt to use as a mixer.
     */
    private static String salt;
    
    /**
     * Generate some salt.
     * Min 8 chars long.
     */
    public static void setSalt(String newsalt) throws SimpleRegistrationException {
	if (newsalt.length() < 8) throw new SimpleRegistrationException("Salt too short");
	salt = newsalt;
    }
    
    /**
     * Return the salt
     * @return String
     */
    public static String getSalt() {
	return SimpleRegistration.salt;
    }
    
    /**
     * My own implementation of hashcode to remain compatibility across versions.
     * @param input
     * @return 
     */
    private static int hashcode(String input) {
	int h = 0;
	int len = input.length();
	for (int i = 0; i < len; i++) {
	    //h = 31 * h + input.charAt(i);
	    h = h + input.charAt(i);
	    System.err.println(h);
	}
	return h;
    }
    
    /**
     * Generate a new code.
     * @param name
     * @return 
     */
    public static String generate(String name) throws SimpleRegistrationException
    {
	if (SimpleRegistration.salt == null)
	    throw new SimpleRegistrationException("Salt not set");
	
	StringBuffer out = new StringBuffer();
	
	name = name.trim();
	name = name.toUpperCase();
	
	String hexString = Integer.toHexString(
		(hashcode(name) << 8) + hashcode(salt)
	);
	char last = '0';
	char current = '0';
	
	for (int n = 0; n < hexString.length(); n++)
	{
	    current = hexString.charAt(n);
	    if (n>0) last = hexString.charAt(n - 1);
	    
	    int charat = Integer.parseInt(Character.toString(current), 16) + Integer.parseInt(Character.toString(last), 16);
	    out.append(HASHTABLE.charAt(charat));
	}
	
	return out.toString();
    }
    
    /**
     * Validate a code.
     * @param name
     * @param codetovalidate
     * @return boolean
     */
    public static boolean validate(String name, String codetovalidate) throws SimpleRegistrationException
    {
	if (name == null) return false;
	if (codetovalidate == null) return false;
	
	name = name.trim();
	codetovalidate = codetovalidate.trim();
	
	if (name.length() == 0) return false;
	if (codetovalidate.length() == 0) return false;
	
	if (codetovalidate.compareTo(generate(name)) == 0)
	    return true;
	
	return false;
    }
    
}

File: php/SimpleRegistration.php
<?php
/**
 * @file
 * Really simple product registration.
 * 
 * Simple registration tool, validates a username and entry code for a user,
 * returns whether it is valid.
 * 
 * Keeping honest people honest :)
 * 
 * @package nppllogbook.logbook
 * @copyright Marcus Povey Consulting, All rights reserved.
 * @license The MIT License (see LICENCE.txt), other licenses available.
 * @author Marcus Povey <marcus@marcus-povey.co.uk>
 * @link http://www.marcus-povey.co.uk
 */
 
 class SimpleRegistration
 {
	 /** Hash table */
	 private static $HASHTABLE = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
	 
	 /** Salt */
	 private static $salt;
	 
	 /**
      * Generate some salt.
      * Min 8 chars long.
      */
	 public static function setSalt($newsalt) {
		 if (strlen($newsalt) < 8) throw new SimpleRegistrationException("Salt too short");
		 self::$salt = $newsalt;
	 }
	 
	 /**
      * Return the salt
      * @return String
      */
	 public static function getSalt() {
		 return self::$salt;
	 }
	 
	 /**
	  * Generate a hashcode (compatible with java String.hashCode())
	  */
	 private static function hashCode($string)
	 {
		$h = 0;
		$len = strlen($string);
	
		for ($i = 0; $i < $len; $i++) {
			
			$h = $h + ord($string[$i]);
		}
		return $h;
	 }
	 
	 /**
      * Generate a new code.
      * @param name
      * @return 
      */
	 public static function generate($name)
	 {
		 if (!self::$salt) 
			throw new SimpleRegistrationException("No salt");
		 
		 $name = trim($name);
		 $name = strtoupper($name);
		 
		 $hexString = base_convert(
			(self::hashCode($name) << 8) + self::hashCode(self::$salt), 10, 16
		 );
		 
		 $last = '0';
		 $current = '0';
		 $out = "";
		 
		 for ($n = 0; $n < strlen($hexString); $n ++)
		 {
			 $current = $hexString[$n];
			 if ($n>0) $last = $hexString[$n-1];
			 
			 $charat = (int)(hexdec($current) + hexdec($last));
			 
			 $out .= self::$HASHTABLE[$charat];
		 }
		 
		 return $out;
	 }
	 
	 /**
      * Validate a code.
      * @param name
      * @param codetovalidate
      * @return boolean
      */
	 public static function validate($name, $codetovalidate)
	 {
		 $name = trim($name);
		 $codetovalidate = trim($codetovalidate);
		 
		 if (!$name) return false;
		 if (!$codetovalidate) return false;
		 
		 if (strcmp($codetovalidate, self::generate($name))==0)
			return true;
			
		 return false;
	 }
 }
 
 class SimpleRegistrationException extends Exception {}
